engine/verification/fetcher/engine.go

```diff
@@ -106,352 +102,209 @@ func (e *Engine) Done() <-chan struct{} {
+	lg.Info().Msg("chunk data pack requested from requester engine")
 }
 
-func blockIsSealed(state protocol.State, headers storage.Headers, blockID flow.Identifier) (bool, *flow.Header, error) {
-	header, err := headers.ByBlockID(blockID)
-	if err != nil {
-		return false, nil, fmt.Errorf("could not get block header by ID: %w", err)
-	}
-
-	lastSealed, err := state.Sealed().Head()
-	if err != nil {
-		return false, nil, fmt.Errorf("could not get last sealed: %w", err)
-	}
-
-	sealed := header.Height <= lastSealed.Height
-	return sealed, header, nil
-}
+// HandleChunkDataPack is called by the chunk requester module everytime a new requested chunk data pack arrives.
+// The chunks are supposed to be deduplicated by the requester.
+// So invocation of this method indicates arrival of a distinct requested chunk.
+func (e *Engine) HandleChunkDataPack(originID flow.Identifier, chunkDataPack *flow.ChunkDataPack, collection *flow.Collection) {
+	lg := e.log.With().
+		Hex("origin_id", logging.ID(originID)).
+		Hex("collection_id", logging.ID(collection.ID())).
+		Hex("chunk_id", logging.ID(chunkDataPack.ChunkID)).
+		Logger()
+	lg.Info().Msg("chunk data pack arrived")
 
-// return agrees and disagrees.
-// agrees are executors who made receipt with the same result as the given result id
-// disagrees are executors who made receipt with different result than the given result id
-func executorsOf(receipts []*flow.ExecutionReceipt, resultID flow.Identifier) ([]flow.Identifier, []flow.Identifier) {
-	agrees := make([]flow.Identifier, 0)
-	disagrees := make([]flow.Identifier, 0)
-	for _, receipt := range receipts {
-		executor := receipt.ExecutorID
-		if receipt.ExecutionResult.ID() == resultID {
-			agrees = append(agrees, executor)
-		} else {
-			disagrees = append(disagrees, executor)
-		}
+	// make sure we still need it
+	status, exists := e.pendingChunks.ByID(chunkDataPack.ChunkID)
+	if !exists {
+		lg.Debug().Msg("could not fetch pending status from mempool, dropping chunk data")
+		return
 	}
-	return agrees, disagrees
-}
-
-// onTimer runs periodically, it goes through all pending chunks, and fetches
-// its chunk data pack.
-// it also retries the chunk data request if the data hasn't been received
-// for a while.
-func (e *Engine) onTimer() {
-	allChunks := e.pendingChunks.All()
 
-	e.log.Debug().Int("total", len(allChunks)).Msg("start processing all pending pendingChunks")
+	lg = lg.With().
+		Uint64("chunk_index", status.Chunk.Index).
+		Hex("block_id", logging.ID(status.Chunk.BlockID)).
+		Hex("result_id", logging.ID(status.ExecutionResultID)).Logger()
 
-	sealed, err := e.state.Sealed().Head()
+	// make sure the chunk data pack is valid
+	err := e.validateChunkDataPack(status.Chunk, originID, chunkDataPack, collection)
 	if err != nil {
-		e.log.Error().Err(err).Msg("could not get last sealed block")
+		// TODO: this can be due to a byzantine behavior
+		lg.Error().Err(err).Msg("could not validate chunk data pack")
 		return
```

Comment:
Despite `fatal` log level, we `return` to make sure that changing the log level does not change the logic.

Reply:
Right, I refactored the code after making the above comment. We decided not to log `fatal` level for invalid chunk data packs, and simply ignore them (and continue to wait for the correct ones). Sorry for the misrepresentation, corrected comment.